cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -177,43 +199,96 @@ private ReportList getReportList(int offset, int limit) throws Exception {
       return new ReportList(offset, limit, reportIdDirs.size(), reportStatuses);
     }
 
+    @POST
+    @Path("/reports/{report-id}/share")
+    public void shareReport(HttpServiceRequest request, HttpServiceResponder responder,
+                                @PathParam("report-id") String reportId) {
+      Location reportIdDir;
+      try {
+        String userName = getUserName(request.getAllHeaders());
+        reportIdDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(userName).append(reportId);
+        if (reportIdDir.exists()) {
+          String shareId = encodeShareId(new ReportIdentifier(userName, reportId));
+          responder.sendJson(new ShareId(shareId));
+        } else {
+          responder.sendError(500, String.format("Invalid report-id %s, report does not exist", reportId));
+          return;
+        }
+      } catch (IOException e) {
+        LOG.error("Failed to get location for report with id {}", reportId, e);
+        responder.sendError(500, String.format("Failed to get location for report with id %s because of error: %s",
+                reportId, e.getMessage()));
+        return;
+      }
+    }
+
+    private String encodeShareId(ReportIdentifier reportIdentifier) {
+      String response = GSON.toJson(reportIdentifier);
```

Comment:
This is not secure, right? We should encrypt the identifier.

Reply:
added encryption, please take a look.